src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/AgreementsApiServiceImpl.scala

```diff
@@ -244,6 +253,44 @@ final case class AgreementsApiServiceImpl(
       eService                         <- catalogManagementService.getEService(agreement.eserviceId)
     } yield (producerTenant.name, consumerTenant.name, consumerTenant, eService)
 
+  def parallelGetTenantsService(agreement: AgreementProcess.Agreement)(implicit
+    contexts: Seq[(String, String)]
+  ): Future[(TenantManagement.Tenant, TenantManagement.Tenant, CatalogManagement.EService)] =
+    Applicative[Future].tuple3(
+      tenantManagementService
+        .getTenant(agreement.consumerId),
+      tenantManagementService.getTenant(agreement.producerId),
+      catalogManagementService.getEService(agreement.eserviceId)
+    )
+
+  def enrichAgreements(
+    agreement: AgreementProcess.Agreement
+  )(implicit contexts: Seq[(String, String)]): Future[CompactAgreements] = for {
+    (consumerTenant, producerTenant, eService) <- parallelGetTenantsService(agreement)
+  } yield CompactAgreements(
+    id = agreement.id,
+    consumer = CompactOrganization(consumerTenant.id, consumerTenant.name),
+    eservice = CompactEService(
+      id = eService.id,
+      name = eService.name,
+      producer = CompactOrganization(producerTenant.id, producerTenant.name)
+    ),
+    upgradable = isUpgradable(agreement.descriptorId.toString, eService.descriptors).getOrElse(false)
+  )
+
+  def isUpgradable(descriptorId: String, descriptors: Seq[CatalogManagement.EServiceDescriptor]): Option[Boolean] =
+    for {
+      currentDescriptor <- descriptors.find(_.id == descriptorId).headOption
```

Comment:
better keeping parameters types strict

```suggestion
  def isUpgradable(descriptorId: UUID, descriptors: Seq[CatalogManagement.EServiceDescriptor]): Option[Boolean] =
    for {
      currentDescriptor <- descriptors.find(_.id.toString == descriptorId).headOption
```